hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastOperandTypes.java

```diff
@@ -0,0 +1,175 @@
+     * disallowing ANY type.
+     */
+    public static SqlSingleOperandTypeChecker notAny(SqlOperandTypeChecker base) {
+        return new NotAny(base);
+    }
+
+    /**
+     * @return the base operand type checker wrapped into a new type checker
+     * disallowing all of the operands to be of NULL type simultaneously.
+     */
+    public static SqlOperandTypeChecker notAllNull(SqlOperandTypeChecker base) {
+        return new NotAllNull(base);
+    }
+
+    private static final class NotAny implements SqlSingleOperandTypeChecker {
+
+        private final SqlOperandTypeChecker base;
+
+        NotAny(SqlOperandTypeChecker base) {
+            this.base = base;
+        }
+
+        @Override
+        public boolean checkOperandTypes(SqlCallBinding binding, boolean throwOnFailure) {
+            if (!base.checkOperandTypes(binding, throwOnFailure)) {
+                return false;
+            }
+
+            for (int i = 0; i < binding.getOperandCount(); ++i) {
+                if (!checkSingleOperandType(binding, binding.operand(i), i, throwOnFailure)) {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        @Override
+        public SqlOperandCountRange getOperandCountRange() {
+            return base.getOperandCountRange();
+        }
+
+        @Override
+        public String getAllowedSignatures(SqlOperator op, String opName) {
+            return base.getAllowedSignatures(op, opName);
+        }
+
+        @Override
+        public Consistency getConsistency() {
+            return base.getConsistency();
+        }
+
+        @Override
+        public boolean isOptional(int i) {
+            return base.isOptional(i);
+        }
+
+        @Override
+        public boolean checkSingleOperandType(SqlCallBinding binding, SqlNode operand, int index, boolean throwOnFailure) {
+            if (typeName(binding.getOperandType(index)) == ANY) {
+                if (throwOnFailure) {
+                    throw binding.newValidationSignatureError();
+                }
+                return false;
+            }
+
+            return true;
+        }
+
+    }
+
+    private static final class NotAllNull implements SqlOperandTypeChecker {
+
+        private final SqlOperandTypeChecker base;
+
+        NotAllNull(SqlOperandTypeChecker base) {
+            this.base = base;
+        }
+
+        @Override
+        public boolean checkOperandTypes(SqlCallBinding binding, boolean throwOnFailure) {
+            boolean seenNonNull = false;
+            for (int i = 0; i < binding.getOperandCount(); ++i) {
+                if (typeName(binding.getOperandType(i)) != NULL) {
+                    seenNonNull = true;
+                    break;
+                }
+            }
+
+            if (!seenNonNull) {
+                if (throwOnFailure) {
+                    throw binding.newValidationSignatureError();
```

Comment:
This error message is misleading. E.g. it will show "cannot apply NULL + NULL, possible combination NUMBER + NUMBER, ...", while in reality `NUMBER + NULL` is possible. Another problem is that this exception will print info about types that are not supported in the scope of this PR, such as temporal types and intervals.
We should provide our own sensible error message here. E.g. this is how PostgreSQL error looks like:
```
ERROR: operator is not unique: unknown + unknown Hint: Could not choose a best candidate operator. You might need to add explicit type casts. Position: 14

```

Reply:
created an issue https://github.com/hazelcast/hazelcast/issues/17301 for 4.2